pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,283 @@
+	d.replicationFactor = promauto.With(r).NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_alertmanager_distributor_replication_factor",
+		Help: "The configured replication factor.",
+	})
+}
+
+func (d *Distributor) starting(ctx context.Context) error {
+	return services.StartManagerAndAwaitHealthy(ctx, d.subservices)
+}
+
+func (d *Distributor) running(ctx context.Context) error {
+	<-ctx.Done()
+	d.requestsInFlight.Wait()
+	return nil
+}
+
+func (d *Distributor) stopping(_ error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), d.subservices)
+}
+
+// IsPathSupported returns true if the given route is currently supported by the Distributor.
+// TODO: This will go away in future after we gradually add support for the entire API.
+func (d *Distributor) IsPathSupported(path string) bool {
+	// API can be found at https://petstore.swagger.io/?url=https://raw.githubusercontent.com/prometheus/alertmanager/master/api/v2/openapi.yaml.
+	return strings.HasSuffix(path, "/alerts") ||
+		strings.HasSuffix(path, "/alerts/groups")
+}
+
+// DistributeRequest shards the writes and returns as soon as the quorum is satisfied.
+// In case of reads, it proxies the request to one of the alertmanagers.
+// DistributeRequest assumes that the caller has verified IsPathSupported return
+// true for the route.
+func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request) {
+	//TODO: Initialise tracing here.
+
+	d.requestsInFlight.Add(1)
+	defer d.requestsInFlight.Done()
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	d.receivedRequests.WithLabelValues(userID).Inc()
+
+	if r.Method == http.MethodGet {
+		d.doRead(userID, w, r)
+	} else {
+		d.doWrite(userID, w, r)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request) {
+	source := util.GetSourceIPsFromOutgoingCtx(r.Context())
+
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+	}
+
+	err = ring.DoBatch(r.Context(), ring.Write, d.alertmanagerRing, []uint32{shardByUser(userID)}, func(am ring.IngesterDesc, _ []int) error {
+		d.amSends.WithLabelValues(am.Addr).Inc()
+
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+		defer cancel()
+		localCtx = user.InjectOrgID(localCtx, userID)
+		if sp := opentracing.SpanFromContext(r.Context()); sp != nil {
+			localCtx = opentracing.ContextWithSpan(localCtx, sp)
+		}
+		// Get clientIP(s) from Context and add it to localCtx
+		localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+
+		// http.Request.Clone() does not do a deep copy, hence copying the body.
+		// https://github.com/golang/go/issues/36095.
+		rClone := r.Clone(r.Context())
```

Comment:
Do we need to close the request? (this is a genuine question) I'm not sure we do if we're using `server.HTTPRequest` as the wrapper for the request.

Reply:
Looks like we don't need to clone the entire request but only need to clone the body (which this clone does not do, it is done separately). I thought of not cloning and creating grpc request manually from 1 request, but the code inside `server.HTTPRequest` has some unexposed methods. I guess it should be fine to copy that small method here and skip this cloning?

Reply:
Yes I will copy that small method and skip cloning.